pkg/supportconfig/auth.go

```diff
@@ -0,0 +1,140 @@
+)
+
+type AuthHandler struct {
+	subjectReview authzv1.SubjectAccessReviewInterface
+	tokenReview   authnv1.TokenReviewInterface
+}
+
+// NewAuthHandler is an Auth handler for specifically the supportconfig endpoint
+func NewAuthHandler(sar authzv1.SubjectAccessReviewInterface, tr authnv1.TokenReviewInterface) *AuthHandler {
+	auth := &AuthHandler{
+		subjectReview: sar,
+		tokenReview:   tr,
+	}
+	return auth
+}
+
+func (a *AuthHandler) Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		userInfo, err := a.authenticate(r)
+		if err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			if err.Error() != unauthenticatedError {
+				logrus.Errorf("error authenticating user %s", err)
+			} else {
+				_, err = w.Write([]byte(err.Error()))
+				if err != nil {
+					logrus.Errorf("error writing response string in auth %s", err.Error())
+				}
+			}
+			// don't go any further if we got an error during auth
+			return
+		}
+		err = a.authorize(r, userInfo)
+		if err != nil {
+			w.WriteHeader(http.StatusForbidden)
+			_, err = w.Write([]byte(err.Error()))
+			if err != nil {
+				logrus.Errorf("error writing response string in auth %s", err.Error())
+			}
+			return
+		}
+		// if we didn't get any errors while authing, it's safe to call the next function
+		next.ServeHTTP(w, r)
+	})
+}
+
+// authenticate issues a token review for the K8s api - basically you need to have `Authorization: Bearer` set like you
+// would for the K8s api. Right now we only have one endpoint so doing authn and authz at the same time makes sense.
+// however, if this changes, this should be broken out into its own middleware
+func (a *AuthHandler) authenticate(r *http.Request) (*authv1.UserInfo, error) {
+	token, err := getTokenFromHeader(r)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := a.tokenReview.Create(r.Context(), &authv1.TokenReview{
+		Spec: authv1.TokenReviewSpec{
+			Token: token,
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// if we aren't authenticated, don't attempt to extract user info
+	if !resp.Status.Authenticated {
+		return nil, fmt.Errorf(unauthenticatedError)
+	}
+	return &resp.Status.User, nil
+}
+
+// authorize checks if a user can get a support config by seeing if they can do "retrieve" on
+// "supportconfigs.managment.cattle.io". As of now, this isn't a real CRD - more a way to do auth in a simple fashion.
+// Intent is that only the admin or a user with these specific perms will be able to get the support configs
+func (a *AuthHandler) authorize(r *http.Request, userInfo *authv1.UserInfo) error {
+	//TODO: Real Resource?
+	resp, err := a.subjectReview.Create(r.Context(), &v1.SubjectAccessReview{
+		Spec: v1.SubjectAccessReviewSpec{
+			ResourceAttributes: &v1.ResourceAttributes{
+				Verb:     "retrieve",
+				Version:  "v1",
+				Resource: "supportconfigs",
+				Group:    "management.cattle.io",
+			},
```

Comment:
How did you test this? It's not possible to create a role for a resource type that doesn't exist

```
$ kubectl create role getsupportconfigs --verb=retrieve --resource=supportconfigs.management.cattle.io
Warning: 'retrieve' is not a standard resource verb
the server doesn't have a resource type "supportconfigs" in group "management.cattle.io"
```

I suggest using the endpoint path instead:

```suggestion
			NonResourceAttributes: &v1.NonResourceAttributes{
				Path: r.URL.Path,,
				Verb: r.Method,
			},
```

which could be used with a clusterrole like this:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: supportconfigs
rules:
- nonResourceURLs: ["/generate/supportconfig"]
  verbs: ["GET"]
```

Reply:
> How did you test this? It's not possible to create a role for a resource type that doesn't exist

Admin binding with a service account (works since admin has *). Technically, full admins are the only ones who need to be able to access this resource, though I really want to leave it open to be able to add/remove granularly in the future. 

One question on the endpoint path- My understanding is that nonResourceURLs are for k8s urls which deal with non-resource endpoints (like `$my_k8s_api_server/healthz`). In this case, we would be requiring that access is given to `$my_k8s_api_server/generate/supportconfig` in addition to the endpoint here (meaning that if this was ever an endpoint for k8s this would give more access than intended). 

Unlikely case, but I still think it's worth asking about. Do I have the correct understanding?

Reply:
That's a good point, using nonResourceURLs here is still somewhat abusing k8s RBAC. If all we need for now is to check whether the user is an admin, could we just check for '*' directly? 
```suggestion
			ResourceAttributes: &v1.ResourceAttributes{
				Verb:     "*",
				Version:  "",
				Resource: "*",
				Group:    "",
			},
```

Reply:
The issue there is that this leaves no room for people to use a custom role - the user that hits this endpoint must be effectively an admin user and nothing else. I'd really like to leave this open to be used in a customized manner. 

I'm kind of thinking that your original idea of a non-resource-url might work as long as we give a prefix that would never be part of the k8s api (like `/rancher-csp-adapter/generate/supportconfig` or maybe that w/o the forward slash).

Reply:
Resolving due to change of goals with this PR.